kmod/core/core.c

```diff
@@ -661,31 +682,22 @@ static int kpatch_write_relocations(struct kpatch_module *kpmod,
        unsigned long core = (unsigned long)kpmod->mod->module_core;
        unsigned long core_size = kpmod->mod->core_size;
 #endif
-	unsigned long src;
+	bool vmlinux = !strcmp(object->name, "vmlinux");
 
 	list_for_each_entry(dynrela, &object->dynrelas, list) {
-		if (!strcmp(object->name, "vmlinux")) {
-			ret = kpatch_verify_symbol_match(dynrela->name,
-							 dynrela->src);
-			if (ret)
-				return ret;
-		} else {
-			/* module, dynrela->src needs to be discovered */
-
-			if (dynrela->external)
-				src = kpatch_find_external_symbol(kpmod,
-								  dynrela->name);
-			else
-				src = kpatch_find_module_symbol(object->mod,
-								dynrela->name);
-
-			if (!src) {
-				pr_err("unable to find symbol '%s'\n",
-				       dynrela->name);
-				return -EINVAL;
-			}
-
-			dynrela->src = src;
+		if (dynrela->external)
+			ret = kpatch_find_external_symbol(kpmod->mod->name,
+							  dynrela->name,
+							  dynrela->sympos,
+							  &dynrela->src);
+		else
+			ret = kpatch_find_object_symbol(vmlinux ? NULL : object->name,
```

Comment:
One minor nit here.  kpatch and klp have a discrepancy between how they specify objname for vmlinux.  For kpatch, it's "vmlinux".  For klp, it's NULL.  Someday we might want to change kpatch to be consistent with klp.

But until then, to reduce confusion, I think kpatch should be consistent with itself.  So here, and for the other caller of this function, I think object->name should be passed directly.  And then kpatch_find_object_symbol() and kpatch_kallsyms()callback should be prepared to deal with "vmlinux" instead of NULL.